File: techniques.py
from mining import *
from os import error

from anvil.region import Region


def standardBranchMining(region: Region, baseDirection: Direction, startingCoords: tuple[int, int, int], numberOfBranchPairs: int, branchLength: int, branchSpacing: int) -> tuple[list[simpleBlock], int]:
    """
    A simple mining technique that consists of a main corridor that has branches coming of it perpendicular to the main corridor.

    Parameters
    ----------
    region
        The region file that the simulation will occur in
    baseDirection
        The direction that the main corridor will be expanding
    startingCoords
        A tuple of the xyz coords that the main corridor will start at
    numberOfBranchPairs
        The number of branches that will stretch out perpendicular to the main corridor
    branchLength
        The length of each branch
    branchSpacing
        The number of blocks in between each pair of branches

    Returns
    -------
    A tuple consisting of all blocks that would be exposed from mining and the number of blocks that would have been mined to replicate the simulation in game.
    """
    if branchSpacing < 2:
        raise error("Branch spacing for branch mining should be at least 2 to avoid duplicates")
        
    def corridorExpansion(coords: tuple[int, int, int]) -> tuple[list[simpleBlock], int]:
        """
        Expands the central corridor based on the branch spacing of the parent function

        Parameters
        ----------
        coords
            The xyz position of where the expansion should start

        Returns
        -------
        A tuple consisting of a list of the blocks exposed and a count of the blocks that would need to be mined.
        """
        blocks = []
        
        # duplicate handeling
        # This is the block that corresponds with the branches
        for y in range(-1, 3):
            blocks.append(simpleBlock(tuple((coords[0], coords[1] + y, coords[2])), getBlock(region, tuple((coords[0], coords[1] + y, coords[2])))))
        # The block after the first one
        for y in range(-1, 3):
            blocks.append(simpleBlock(dirCoords(baseDirection, tuple((coords[0], coords[1] + y, coords[2]))), getBlock(region, dirCoords(baseDirection, tuple((coords[0], coords[1] + y, coords[2]))))))
        # Final block of the corridor
        for y in range(-1, 3):
            blocks.append(simpleBlock(dirCoords(baseDirection, tuple((coords[0], coords[1] + y, coords[2])), branchSpacing), getBlock(region, dirCoords(baseDirection, tuple((coords[0], coords[1] + y, coords[2])), branchSpacing))))
        
        for n in range(2, branchSpacing):
            blocks += twoByOne(region, baseDirection, dirCoords(baseDirection, coords, n))

        return tuple[blocks, 2 + (2 * branchSpacing)]

    def branch(direction: Direction, coords: tuple[int, int, int]) -> tuple[list[simpleBlock], int]:
        """
        Simulates mining a branch in a direction

        Parameters
        ----------
        direction
            The direction that the branch should expand in
        coords
            The xyz position that the branch should start at

        Returns
        -------
        A tuple consisting of a list of the blocks exposed and a count of the blocks that would need to be mined.
        """
        blocks = []
        mined = 0
        for n in range(branchLength):
            blocks += twoByOne(region, direction, dirCoords(direction, coords, n))
            mined += 2

        blocks += twoByOneEnd(region, direction, dirCoords(direction, coords, branchLength - 1))
        return tuple[blocks, mined]

    sumBlocks = []
    mined = 0
    branchDirection1, branchDirection2 = (Direction.NORTH, Direction.SOUTH) if baseDirection is Direction.EAST or baseDirection is Direction.WEST else Direction.EAST, Direction.WEST

    (tempBlocks, tempmined) = branch(branchDirection1, dirCoords(branchDirection1, startingCoords))
    sumBlocks += tempBlocks
    mined += tempmined
    (tempBlocks, tempmined) = branch(branchDirection2, dirCoords(branchDirection1, startingCoords))
    sumBlocks += tempBlocks
    mined += tempmined

    for n in range(numberOfBranchPairs - 1):
        (tempBlocks, tempmined) = corridorExpansion(dirCoords(baseDirection, startingCoords, n * 13))
        sumBlocks += tempBlocks
        mined += tempmined
        (tempBlocks, tempmined) = branch(branchDirection1, dirCoords(branchDirection1, startingCoords, n * 13))
        sumBlocks += tempBlocks
        mined += tempmined
        (tempBlocks, tempmined) = branch(branchDirection2, dirCoords(branchDirection1, startingCoords, n * 13))
        sumBlocks += tempBlocks
        mined += tempmined

    return tuple[sumBlocks, mined]

def branchWithPokeHoles(region: Region, baseDirection: Direction, startingCoords: tuple[int, int, int], numberOfBranchPairs: int, pokesPerBranch: int, pokeSpacing: int, branchSpacing: int) -> tuple[list[simpleBlock], int]:
    """
    
    """
    if branchSpacing < 10:
        raise error("Branch spacing for branch mining with poke holes should be at least 10 to avoid duplicates")
    if pokeSpacing < 2:
        raise error("Poke spacing should be at least 2 to avoid duplicates")

    def corridorExpansion(coords: tuple[int, int, int]):
        """
        
        """
        blocks = []
        
        # duplicate handeling
        # This is the block that corresponds with the branches
        for y in range(-1, 3):
            blocks.append(simpleBlock(tuple((coords[0], coords[1] + y, coords[2])), getBlock(region, tuple((coords[0], coords[1] + y, coords[2])))))
        # The block after the first one
        for y in range(-1, 3):
            blocks.append(simpleBlock(dirCoords(baseDirection, tuple((coords[0], coords[1] + y, coords[2])), 1), getBlock(region, dirCoords(baseDirection, tuple((coords[0], coords[1] + y, coords[2]))))))
        # Final block of the corridor
        for y in range(-1, 3):
            blocks.append(simpleBlock(dirCoords(baseDirection, tuple((coords[0], coords[1] + y, coords[2])), branchSpacing), getBlock(region, dirCoords(baseDirection, tuple((coords[0], coords[1] + y, coords[2])), branchSpacing))))
        
        for n in range(2, 12):
            blocks += twoByOne(region, baseDirection, dirCoords(baseDirection, coords, n))

        return tuple((blocks, 2 + (2 * branchSpacing)))

    def branch(direction: Direction, coords: tuple[int, int, int]):
        """
        
        """
        blocks = []
        mined = 0
        for n in range(pokesPerBranch):
            coords = dirCoords(direction, coords, n * pokeSpacing) # Offset for the starting postion of each poke section
            # Base tunnel of branch
            for n in range(pokeSpacing):
                blocks += twoByOne(region, direction, dirCoords(direction, coords, n))

            # Poke start
            pokeCoords = list(coords)
            pokeCoords[1] += 1 # Raise the y by 1 for the pokes
            pokeCoords = dirCoords(direction, tuple(pokeCoords), pokeSpacing)
            # Gets directions perpendicular to the base tunnel direction
            if direction == Direction.EAST or direction == Direction.WEST:
                pokeDirection1, pokeDirection2 = Direction.NORTH, Direction.SOUTH  
            else:
                pokeDirection1, pokeDirection2 = Direction.EAST, Direction.WEST
            
            # The first side poke
            blocks += pokeStart(region, pokeDirection1, dirCoords(pokeDirection1, pokeCoords))
            blocks += oneByOne(region, pokeDirection1, dirCoords(pokeDirection1, pokeCoords, 2))
            blocks += oneByOne(region, pokeDirection1, dirCoords(pokeDirection1, pokeCoords, 3))
            blocks += oneByOne(region, pokeDirection1, dirCoords(pokeDirection1, pokeCoords, 4))
            blocks += oneByOneEnd(region, pokeDirection1, dirCoords(pokeDirection1, pokeCoords, 5))

            # Second side poke
            blocks += pokeStart(region, pokeDirection2, dirCoords(pokeDirection2, pokeCoords))
            blocks += oneByOne(region, pokeDirection2, dirCoords(pokeDirection2, pokeCoords, 2))
            blocks += oneByOne(region, pokeDirection2, dirCoords(pokeDirection2, pokeCoords, 3))
            blocks += oneByOne(region, pokeDirection2, dirCoords(pokeDirection2, pokeCoords, 4))
            blocks += oneByOneEnd(region, pokeDirection2, dirCoords(pokeDirection2, pokeCoords, 5))
            mined += 10 + (2 * pokeSpacing)

        blocks += twoByOneEnd(region, direction, dirCoords(direction, coords, (pokeSpacing * pokesPerBranch) - 1))
        return tuple[blocks, mined]

    sumBlocks = []
    mined = 0
    branchDirection1, branchDirection2 = (Direction.NORTH, Direction.SOUTH) if baseDirection is Direction.EAST or baseDirection is Direction.WEST else Direction.EAST, Direction.WEST

    (tempBlocks, tempmined) = branch(branchDirection1, dirCoords(branchDirection1, startingCoords))
    sumBlocks += tempBlocks
    mined += tempmined
    (tempBlocks, tempmined) = branch(branchDirection2, dirCoords(branchDirection1, startingCoords))
    sumBlocks += tempBlocks
    mined += tempmined

    for n in range(numberOfBranchPairs - 1):
        (tempBlocks, tempmined) = corridorExpansion(dirCoords(baseDirection, startingCoords, n * 13))
        sumBlocks += tempBlocks
        mined += tempmined
        (tempBlocks, tempmined) = branch(branchDirection1, dirCoords(branchDirection1, startingCoords, n * 13))
        sumBlocks += tempBlocks
        mined += tempmined
        (tempBlocks, tempmined) = branch(branchDirection2, dirCoords(branchDirection1, startingCoords, n * 13))
        sumBlocks += tempBlocks
        mined += tempmined

    return tuple[sumBlocks, mined]

File: src/techniques.rs
use std::sync::mpsc::Sender;

use mvp_anvil::region::Region;

use crate::{ProgramStatus, mining::*};

#[derive(Clone)]
pub enum Technique {
    Branch,
    BranchWithPoke,
}

impl Technique {
    pub fn name(self) -> String {
        match self {
            Self::Branch => String::from("branch"),
            Self::BranchWithPoke => String::from("poke"),
        }
    }

    pub fn iterable() -> Vec<String> {
        return vec!["Branch", "Branch with Poke Holes"].iter().map(|f| f.to_string()).collect();
    }

    pub fn from_string(text: String) -> Technique {
        match text.as_str() {
            "Branch" => Technique::Branch,
            "Branch with Poke Holes" => Technique::BranchWithPoke,
            _ => unreachable!("only techs")
        }
    }
}

pub fn branch_mining(
    region: &Region,
    base_direction: &Direction,
    starting_coords: (i32, i32, i32),
    branch_pair_count: i32,
    branch_length: i32,
    branch_spacing: i32,
    id: u32,
    sender: Sender<ProgramStatus>,
) -> (Vec<SimpleBlock>, u32, u32) {
    if branch_spacing < 2 {
        panic!("Branch spacing should be at least two to avoid duplicates")
    }

    fn expand_corridor(
        region: &Region,
        direction: &Direction,
        branch_spacing: i32,
        coords: (i32, i32, i32),
    ) -> (Vec<SimpleBlock>, u32, u32) {
        let mut results = (Vec::new(), 0, 0);
        // The sides of the corridor that coincide with a branch along with the sides in front and behind. For these we don't consider the side blocks that would be exposed.
        for y in -1..3 {
            // Slice that coincides
            results.0.push(SimpleBlock::new(
                (coords.0, coords.1 + y, coords.2),
                get_block(region, (coords.0, coords.1 + y, coords.2)),
            ));
            results.1 += 2;
            results.2 += 4;
            // Slice after
            results.0.push(SimpleBlock::new(
                shift_coords(direction, (coords.0, coords.1 + y, coords.2), 1),
                get_block(
                    region,
                    shift_coords(direction, (coords.0, coords.1 + y, coords.2), 1),
                ),
            ));
            results.1 += 2;
            results.2 += 4;
            // Slice before
            results.0.push(SimpleBlock::new(
                shift_coords(direction, (coords.0, coords.1 + y, coords.2), branch_spacing),
                get_block(
                    region,
                    shift_coords(direction, (coords.0, coords.1 + y, coords.2), branch_spacing),
                ),
            ));
            results.1 += 2;
            results.2 += 4;
        }

        let mut res = two_by_one_length(
            region,
            direction,
            shift_coords(direction, coords, 2),
            branch_spacing - 3,
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
        return results;
    }

    fn branch(
        region: &Region,
        branch_length: i32,
        direction: &Direction,
        coords: (i32, i32, i32),
    ) -> (Vec<SimpleBlock>, u32, u32) {
        let mut results = (Vec::new(), 0, 0);

        let mut res = two_by_one_length(region, direction, coords, branch_length);
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;

        let mut res = two_by_one_end(
            region,
            direction,
            shift_coords(direction, coords, branch_length - 1),
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;

        return results;
    }

    let mut results = (Vec::new(), 0, 0);
    let (branch_direction1, branch_direction2) =
        if base_direction == &Direction::East || base_direction == &Direction::West {
            (&Direction::North, &Direction::South)
        } else {
            (&Direction::East, &Direction::West)
        };
        sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Branch 1/{}", branch_pair_count * 2), 0, 0, 0, 0)).unwrap();
    let mut res = branch(
        region,
        branch_length,
        branch_direction1,
        shift_coords(branch_direction1, starting_coords, 1),
    );
    results.0.append(&mut res.0);
    results.1 += res.1;
    results.2 += res.2;
    sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Branch 2/{}", branch_pair_count * 2), results.1, results.2, 0, 0)).unwrap();
    let mut res = branch(
        region,
        branch_length,
        branch_direction2,
        shift_coords(branch_direction2, starting_coords, 1),
    );
    results.0.append(&mut res.0);
    results.1 += res.1;
    results.2 += res.2;
    for n in 0..branch_pair_count - 1 {
        sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Corridor"), results.1, results.2, 0, 0)).unwrap();
        let mut res = expand_corridor(
            region,
            base_direction,
            branch_spacing,
            shift_coords(base_direction, starting_coords, n * branch_spacing),
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
        sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Branch {}/{}", 2 + (n * 2), branch_pair_count * 2), results.1, results.2, 0, 0)).unwrap();
        let mut res = branch(
            region,
            branch_length,
            branch_direction1,
            shift_coords(
                base_direction,
                shift_coords(branch_direction1, starting_coords, 1),
                (n + 1) * branch_spacing,
            ),
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
        sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Branch {}/{}", 2 + (n * 2) + 1, branch_pair_count * 2), results.1, results.2, 0, 0)).unwrap();
        let mut res = branch(
            region,
            branch_length,
            branch_direction2,
            shift_coords(
                base_direction,
                shift_coords(branch_direction2, starting_coords, 1),
                (n + 1) * branch_spacing,
            ),
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
    }

    return results;
}

pub fn branch_mining_with_poke_holes(
    region: &Region,
    base_direction: &Direction,
    starting_coords: (i32, i32, i32),
    branch_pair_count: i32,
    pokes_per_branch: i32,
    poke_spacing: i32,
    branch_spacing: i32,
    id: u32,
    sender: Sender<ProgramStatus>,
) -> (Vec<SimpleBlock>, u32, u32) {
    fn expand_corridor(
        region: &Region,
        direction: &Direction,
        branch_spacing: i32,
        coords: (i32, i32, i32),
    ) -> (Vec<SimpleBlock>, u32, u32) {
        let mut results = (Vec::new(), 0, 0);
        // The sides of the corridor that coincide with a branch along with the sides in front and behind. For these we don't consider the side blocks that would be exposed.
        for y in -1..3 {
            // Slice that coincides
            results.0.push(SimpleBlock::new(
                (coords.0, coords.1 + y, coords.2),
                get_block(region, (coords.0, coords.1 + y, coords.2)),
            ));
            results.1 += 2;
            results.2 += 4;
            // Slice after
            results.0.push(SimpleBlock::new(
                shift_coords(direction, (coords.0, coords.1 + y, coords.2), 1),
                get_block(
                    region,
                    shift_coords(direction, (coords.0, coords.1 + y, coords.2), 1),
                ),
            ));
            results.1 += 2;
            results.2 += 4;
            // Slice before
            results.0.push(SimpleBlock::new(
                shift_coords(direction, (coords.0, coords.1 + y, coords.2), branch_spacing),
                get_block(
                    region,
                    shift_coords(direction, (coords.0, coords.1 + y, coords.2), branch_spacing),
                ),
            ));
            results.1 += 2;
            results.2 += 4;
        }

        let mut res = two_by_one_length(
            region,
            direction,
            shift_coords(direction, coords, 2),
            branch_spacing - 3,
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
        return results;
    }

    fn branch(
        region: &Region,
        pokes_per_branch: i32,
        poke_spacing: i32,
        direction: &Direction,
        coords: (i32, i32, i32),
    ) -> (Vec<SimpleBlock>, u32, u32) {
        let mut results = (Vec::new(), 0, 0);
        for n in 0..pokes_per_branch {
            let (poke_direction1, poke_direction2) =
                if direction == &Direction::East || direction == &Direction::West {
                    (&Direction::North, &Direction::South)
                } else {
                    (&Direction::East, &Direction::West)
                };
            let coords = shift_coords(direction, coords, n * poke_spacing - 1);
            let mut res = poke(
                region,
                poke_direction1,
                shift_coords(poke_direction1, coords, 1),
                5,
            );
            results.0.append(&mut res.0);
            results.1 += res.1;
            results.2 += res.2;
            let mut res = poke(
                region,
                poke_direction2,
                shift_coords(poke_direction2, coords, 1),
                5,
            );
            results.0.append(&mut res.0);
            results.1 += res.1;
            results.2 += res.2;
        }
        let mut res = two_by_one_length(region, direction, coords, poke_spacing * pokes_per_branch);
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
        let mut res = two_by_one_end(
            region,
            direction,
            shift_coords(direction, coords, poke_spacing * pokes_per_branch),
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
        return results;
    }

    let mut results = (Vec::new(), 0, 0);
    let (branch_direction1, branch_direction2) =
        if base_direction == &Direction::East || base_direction == &Direction::West {
            (&Direction::North, &Direction::South)
        } else {
            (&Direction::East, &Direction::West)
        };
    sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Branch 1/{}", branch_pair_count * 2), 0, 0, 0, 0)).unwrap();
    let mut res = branch(
        region,
        pokes_per_branch,
        poke_spacing,
        branch_direction1,
        shift_coords(branch_direction1, starting_coords, 1),
    );
    results.0.append(&mut res.0);
    results.1 += res.1;
    results.2 += res.2;
    sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Branch 2/{}", branch_pair_count * 2), results.1, results.2, 0, 0)).unwrap();
    let mut res = branch(
        region,
        pokes_per_branch,
        poke_spacing,
        branch_direction2,
        shift_coords(branch_direction2, starting_coords, 1),
    );
    results.0.append(&mut res.0);
    results.1 += res.1;
    results.2 += res.2;
    for n in 0..branch_pair_count - 1 {
        sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Corridor"), results.1, results.2, 0, 0)).unwrap();
        let mut res = expand_corridor(
            region,
            base_direction,
            branch_spacing,
            shift_coords(base_direction, starting_coords, n * branch_spacing),
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
        sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Branch {}/{}", 2 + (n * 2), branch_pair_count * 2), results.1, results.2, 0, 0)).unwrap();
        let mut res = branch(
            region,
            pokes_per_branch,
            poke_spacing,
            branch_direction1,
            shift_coords(
                base_direction,
                shift_coords(branch_direction1, starting_coords, 1),
                (n + 1) * branch_spacing,
            ),
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
        sender.send(ProgramStatus::UpdateSim(id, format!("Simulating Branch {}/{}", 2 + (n * 2) + 1, branch_pair_count * 2), results.1, results.2, 0, 0)).unwrap();
        let mut res = branch(
            region,
            pokes_per_branch,
            poke_spacing,
            branch_direction2,
            shift_coords(
                base_direction,
                shift_coords(branch_direction2, starting_coords, 1),
                (n + 1) * branch_spacing,
            ),
        );
        results.0.append(&mut res.0);
        results.1 += res.1;
        results.2 += res.2;
    }

    return results;
}
